tokio/src/sync/notify.rs

```diff
@@ -766,6 +776,17 @@ impl Notified<'_> {
                         return Poll::Ready(());
                     }
 
+                    // Safety: the waiter is still not inserted
+                    let initial_notify_waiters_calls =
+                        unsafe { (*waiter.get()).notify_waiters_calls };
+
+                    // Optimistically check if notify_waiters has been called
+                    // after the future was created.
+                    if get_num_notify_waiters_calls(curr) != initial_notify_waiters_calls {
+                        *state = Done;
+                        return Poll::Ready(());
+                    }
```

Comment:
This additional check is basically free and should help to avoid waiting for the lock.